lib/events/gcssessions/gcshandler.go

```diff
@@ -0,0 +1,319 @@
+	if len(credentialsPathParamString) > 0 {
+		cfg.CredentialsPath = credentialsPathParamString
+	}
+
+	projectIDParamString := url.Query().Get(projectID)
+	if projectIDParamString == "" {
+		return trace.BadParameter("parameter %s with value '%s' is invalid",
+			projectID, projectIDParamString)
+	} else {
+		cfg.ProjectID = projectIDParamString
+	}
+
+	if url.Host == "" {
+		return trace.BadParameter("host should be set to the bucket name for recording storage")
+	} else {
+		cfg.Bucket = url.Host
+	}
+
+	return nil
+}
+
+// DefaultNewHandler returns a new handler with default GCS client settings derived from the config
+func DefaultNewHandler(cfg Config) (*Handler, error) {
+
+	var args []option.ClientOption
+	if len(cfg.Endpoint) != 0 {
+		args = append(args, option.WithoutAuthentication(), option.WithEndpoint(cfg.Endpoint), option.WithGRPCDialOption(grpc.WithInsecure()))
+	} else if len(cfg.CredentialsPath) != 0 {
+		args = append(args, option.WithCredentialsFile(cfg.CredentialsPath))
+	}
+
+	ctx, cancelFuc := context.WithCancel(context.Background())
+
+	client, err := storage.NewClient(ctx, args...)
+
+	if err != nil {
+		return nil, trace.BadParameter("error creating GCS gcsClient; error %v", convertGCSError(err))
+	}
+
+	return NewHandler(ctx, cancelFuc, cfg, client)
+}
+
+// NewHandler returns a new handler with specific context, cancelFunc, and client
+func NewHandler(ctx context.Context, cancelFunc context.CancelFunc, cfg Config, client *storage.Client) (*Handler, error) {
+	h := &Handler{
+		Entry: log.WithFields(log.Fields{
+			trace.Component: teleport.Component(teleport.SchemeGCS),
+		}),
+		Config:        cfg,
+		gcsClient:     client,
+		clientContext: ctx,
+		clientCancel:  cancelFunc,
+	}
+	start := time.Now()
+	h.Infof("Setting up bucket %q, sessions path %q.", h.Bucket, h.Path)
+	if err := h.ensureBucket(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	h.WithFields(log.Fields{"duration": time.Now().Sub(start)}).Infof("Setup bucket %q completed.", h.Bucket)
+	return h, nil
+}
+
+// Handler handles upload and downloads to GCS object storage
+type Handler struct {
+	// Config is handler configuration
+	Config
+	// Entry is a logging entry
+	*log.Entry
+	// gcsClient is the google cloud storage client used for persistence
+	gcsClient *storage.Client
+	// clientContext is used for non-request operations and cleanup
+	clientContext context.Context
+	// clientCancel is a function that will cancel the clientContext
+	clientCancel context.CancelFunc
+}
+
+// Closer releases connection and resources associated with log if any
+func (h *Handler) Close() error {
+	h.clientCancel()
+	return h.gcsClient.Close()
+}
+
+// Upload uploads object to GCS bucket, reads the contents of the object from reader
+// and returns the target GCS bucket path in case of successful upload.
+func (h *Handler) Upload(ctx context.Context, sessionID session.ID, reader io.Reader) (string, error) {
```

Comment:
What happens if the session ID is the same as an existing session recording? Will it be clobbered?